File: tools/ods/internal/impersonate/impersonate.py
"""Runs inside a customer api-server pod, fed to `python -` on stdin.

Takes one argv: a JSON request blob. Writes one sentinel-prefixed JSON line
to stdout. Embedded into the ods binary by impersonate.go.
"""

import json
import os
import secrets
import sys
from datetime import datetime, timedelta, timezone
from typing import Any

import psycopg2
import redis
from psycopg2 import sql

# Deliberately imports nothing from onyx: this runs against whatever image the
# customer is on, and onyx module paths move between releases. Only the env vars,
# the key prefix, and the token JSON shape are relied on, and those are stable.
RESULT_SENTINEL = "__ONYX_IMPERSONATE_RESULT__"
AUTH_KEY_PREFIX = "fastapi_users_token:"
SCAN_COUNT = 1000


def _schema() -> str:
    return os.environ.get("POSTGRES_DEFAULT_SCHEMA") or "public"


def _redis_client() -> "redis.Redis[Any]":
    kwargs: dict[str, Any] = {
        "host": os.environ.get("REDIS_HOST") or "localhost",
        "port": int(os.environ.get("REDIS_PORT") or 6379),
        "db": int(os.environ.get("REDIS_DB_NUMBER") or 0),
        "password": os.environ.get("REDIS_PASSWORD") or None,
        "decode_responses": True,
    }
    if (os.environ.get("REDIS_SSL") or "").lower() == "true":
        kwargs["ssl"] = True
        kwargs["ssl_cert_reqs"] = os.environ.get("REDIS_SSL_CERT_REQS") or "none"
        ca_certs = os.environ.get("REDIS_SSL_CA_CERTS")
        if ca_certs:
            kwargs["ssl_ca_certs"] = ca_certs
    return redis.Redis(**kwargs)


def _pg_connect() -> Any:
    return psycopg2.connect(
        host=os.environ.get("POSTGRES_HOST") or "127.0.0.1",
        port=os.environ.get("POSTGRES_PORT") or "5432",
        user=os.environ.get("POSTGRES_USER") or "postgres",
        password=os.environ.get("POSTGRES_PASSWORD") or "",
        dbname=os.environ.get("POSTGRES_DB") or "postgres",
    )


def _classify(payload: dict[str, Any]) -> str:
    """Mirrors onyx.auth.session_tokens without importing it, for both formats."""
    if payload.get("logged_out_at"):
        return "terminated"
    if not payload.get("sub"):
        return "malformed"

    expires_at = payload.get("expires_at")
    if not expires_at:
        # Pre-July-2026 tokens carry no expiry; the Redis TTL is their lifetime.
        return "active"
    try:
        if datetime.fromisoformat(expires_at) <= datetime.now(timezone.utc):
            return "expired"
    except ValueError:
        return "malformed"
    return "active"


def _scan_sessions(redis_client: "redis.Redis[Any]") -> list[dict[str, Any]]:
    """Reads every session token, with its status, remaining TTL, and payload."""
    sessions: list[dict[str, Any]] = []
    for key in redis_client.scan_iter(AUTH_KEY_PREFIX + "*", count=SCAN_COUNT):
        raw_value = redis_client.get(key)
        if raw_value is None:
            continue

        try:
            payload = json.loads(raw_value)
        except json.JSONDecodeError:
            payload = {}

        sessions.append(
            {
                "token": key[len(AUTH_KEY_PREFIX) :],
                "sub": payload.get("sub"),
                "status": _classify(payload),
                "ttl_seconds": redis_client.ttl(key),
                "expires_at": payload.get("expires_at"),
                "impersonated_by": payload.get("impersonated_by"),
            }
        )
    return sessions


def _resolve_emails(subs: list[str]) -> dict[str, str]:
    if not subs:
        return {}

    query = sql.SQL('SELECT id, email FROM {}."user" WHERE id::text = ANY(%s)').format(
        sql.Identifier(_schema())
    )
    with _pg_connect() as conn, conn.cursor() as cur:
        cur.execute(query, (subs,))
        return {str(row[0]): row[1] for row in cur.fetchall()}


def _lookup_user(email: str) -> tuple[str, str, str] | None:
    """Returns (id, email, role), matched case-insensitively like get_user_by_email."""
    query = sql.SQL(
        'SELECT id, email, role FROM {}."user" WHERE lower(email) = lower(%s)'
    ).format(sql.Identifier(_schema()))
    with _pg_connect() as conn, conn.cursor() as cur:
        cur.execute(query, (email,))
        row = cur.fetchone()
        return (str(row[0]), row[1], str(row[2])) if row else None


def _mint(request: dict[str, Any]) -> dict[str, Any]:
    email = str(request["email"])
    ttl_seconds = int(request["ttl_seconds"])

    found = _lookup_user(email)
    if found is None:
        return {"ok": False, "error": "No user found with email " + email}
    user_id, actual_email, role = found

    now = datetime.now(timezone.utc)
    expires_at = now + timedelta(seconds=ttl_seconds)
    # A superset of both formats. Readers before July 2026 take "sub" and ignore
    # the rest; later ones parse the timestamps and drop the unknown marker.
    payload = {
        "sub": user_id,
        "tenant_id": _schema(),
        "issued_at": now.isoformat(),
        "expires_at": expires_at.isoformat(),
        "impersonated_by": request["operator"],
    }

    token = secrets.token_urlsafe()
    # TTL matches the logical expiry exactly. Newer deployments would add an hour
    # of grace on top; skipping it keeps the token dead on time on older ones.
    _redis_client().set(AUTH_KEY_PREFIX + token, json.dumps(payload), ex=ttl_seconds)
    return {
        "ok": True,
        "token": token,
        "email": actual_email,
        "user_id": user_id,
        "role": role,
        "expires_at": expires_at.isoformat(),
    }


def _list(_request: dict[str, Any]) -> dict[str, Any]:
    sessions = _scan_sessions(_redis_client())
    emails = _resolve_emails([s["sub"] for s in sessions if s["sub"]])
    for session in sessions:
        session["email"] = emails.get(session["sub"] or "", "<unknown>")
    return {"ok": True, "sessions": sessions}


def _revoke(request: dict[str, Any]) -> dict[str, Any]:
    token = request.get("token")
    include_real_sessions = bool(request.get("include_real_sessions"))

    redis_client = _redis_client()
    if token:
        deleted = redis_client.delete(AUTH_KEY_PREFIX + str(token))
        return {"ok": True, "deleted": int(deleted), "tokens": [token]}

    email = str(request["email"])
    found = _lookup_user(email)
    if found is None:
        return {"ok": False, "error": "No user found with email " + email}
    user_id = found[0]

    revoked: list[str] = []
    for session in _scan_sessions(redis_client):
        if session["sub"] != user_id:
            continue
        if not include_real_sessions and not session["impersonated_by"]:
            continue
        redis_client.delete(AUTH_KEY_PREFIX + session["token"])
        revoked.append(session["token"])

    return {"ok": True, "deleted": len(revoked), "tokens": revoked}


def main() -> int:
    request = json.loads(sys.argv[1])
    command = request["command"]

    handlers = {"mint": _mint, "list": _list, "revoke": _revoke}
    try:
        result = handlers[command](request)
    except Exception as e:
        result = {"ok": False, "error": type(e).__name__ + ": " + str(e)}

    print(RESULT_SENTINEL + json.dumps(result))
    return 0 if result.get("ok") else 1


sys.exit(main())

File: tools/ods/internal/impersonate/impersonate.go
// Package impersonate mints, lists, and revokes session tokens on a single-tenant
// deployment by running an embedded Python payload inside the api-server pod.
//
// The payload imports nothing from onyx: customers run older images whose module
// paths differ from main. It uses the pod's env vars with redis and psycopg2, and
// writes a token that satisfies both session formats -- readers before July 2026
// take "sub" and ignore the rest, later ones parse issued_at/expires_at.
package impersonate

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/onyx-dot-app/onyx/tools/ods/internal/kube"
)

//go:embed impersonate.py
var embeddedScript string

// resultSentinel prefixes the payload's single result line, so onyx logging on
// stdout around it is ignored.
const resultSentinel = "__ONYX_IMPERSONATE_RESULT__"

// APIServerPod is the pod substring the payload runs on.
const APIServerPod = "api-server"

// Session is one entry in the pod's Redis session store.
type Session struct {
	Token          string `json:"token"`
	Sub            string `json:"sub"`
	Email          string `json:"email"`
	Status         string `json:"status"`
	TTLSeconds     int    `json:"ttl_seconds"`
	ExpiresAt      string `json:"expires_at"`
	ImpersonatedBy string `json:"impersonated_by"`
}

// Result is the payload's response. Only the fields for the request's command are set.
type Result struct {
	OK    bool   `json:"ok"`
	Error string `json:"error"`

	// mint
	Token     string `json:"token"`
	Email     string `json:"email"`
	UserID    string `json:"user_id"`
	Role      string `json:"role"`
	ExpiresAt string `json:"expires_at"`

	// list
	Sessions []Session `json:"sessions"`

	// revoke
	Deleted int      `json:"deleted"`
	Tokens  []string `json:"tokens"`
}

// run ships the payload to the pod and parses its sentinel line.
func run(c *kube.Cluster, pod string, request map[string]any) (*Result, error) {
	encoded, err := json.Marshal(request)
	if err != nil {
		return nil, fmt.Errorf("failed to encode request: %w", err)
	}

	stdout, err := c.ExecOnPodWithStdin(pod, embeddedScript, "python", "-", string(encoded))
	if err != nil {
		return nil, err
	}

	for _, line := range strings.Split(stdout, "\n") {
		if !strings.HasPrefix(line, resultSentinel) {
			continue
		}
		var result Result
		if err := json.Unmarshal([]byte(strings.TrimPrefix(line, resultSentinel)), &result); err != nil {
			return nil, fmt.Errorf("failed to decode result: %w", err)
		}
		if !result.OK {
			return nil, fmt.Errorf("%s", result.Error)
		}
		return &result, nil
	}

	return nil, fmt.Errorf("the api-server pod returned no result:\n%s", strings.TrimSpace(stdout))
}

// Mint writes a new session token for the given user and returns its cookie value.
func Mint(c *kube.Cluster, pod, email string, ttlSeconds int, operator string) (*Result, error) {
	return run(c, pod, map[string]any{
		"command":     "mint",
		"email":       email,
		"ttl_seconds": ttlSeconds,
		"operator":    operator,
	})
}

// List returns every session token currently in Redis.
func List(c *kube.Cluster, pod string) (*Result, error) {
	return run(c, pod, map[string]any{"command": "list"})
}

// Revoke deletes a single token, or every impersonation session for an email.
// Set includeRealSessions to also drop the user's own logins.
func Revoke(c *kube.Cluster, pod, email, token string, includeRealSessions bool) (*Result, error) {
	return run(c, pod, map[string]any{
		"command":               "revoke",
		"email":                 email,
		"token":                 token,
		"include_real_sessions": includeRealSessions,
	})
}
